src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Types.jl

```diff
@@ -44,6 +44,7 @@ defined by ideal(y, x)
     A.X = X
     if is_reduced
       A.Xred = X
+      A.__attrs = copy(X.__attrs)
```

Comment:
This piece of code relies on internals of how attributes are stored in an object (which are in general not true).

Furthermore, it seems dangerous (or even incorrect) to just copy all attributes, without checking if the attribute is even applicable to the other type.
The better way would be to do something along the following:
```suggestion
      for attr_key in [:some_key, :some_other_key]
        if has_attribute(X, attr_key)
          set_attribute!(A, get_attribute(X, attr_key))
        end
      end
```
(same for the other places where this pattern is repeated in this PR)

Reply:
> This piece of code relies on internals of how attributes are stored in an object (which are in general not true).

```julia
Indeed, this runs into failures in the CI: https://github.com/oscar-system/Oscar.jl/actions/runs/16318334996/job/46089748379?pr=5118#step:10:1605
      From worker 3:	issue 3107: Error During Test at /home/oscarci-tester/ssd-data/ssd-runner-21/_work/Oscar.jl/Oscar.jl/test/Modules/UngradedModules.jl:1333
      From worker 3:	  Got exception outside of a @test
      From worker 3:	  UndefRefError: access to undefined reference
      From worker 3:	  Stacktrace:
      From worker 3:	    [1] getproperty
      From worker 3:	      @ ./Base.jl:49 [inlined]
      From worker 3:	    [2] ProjectiveAlgebraicSet(X::ProjectiveScheme{fpField, MPolyQuoRing{MPolyDecRingElem{fpFieldElem, fpMPolyRingElem}}}; is_reduced::Bool, check::Bool)
      From worker 3:	      @ Oscar ~/ssd-data/ssd-runner-21/_work/Oscar.jl/Oscar.jl/src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Types.jl:10
      From worker 3:	    [3] ProjectiveAlgebraicSet
      From worker 3:	      @ ~/ssd-data/ssd-runner-21/_work/Oscar.jl/Oscar.jl/src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Types.jl:4 [inlined]
      From worker 3:	    [4] #algebraic_set#5796
      From worker 3:	      @ ~/ssd-data/ssd-runner-21/_work/Oscar.jl/Oscar.jl/src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Constructors.jl:28 [inlined]
      From worker 3:	    [5] algebraic_set
      From worker 3:	      @ ~/ssd-data/ssd-runner-21/_work/Oscar.jl/Oscar.jl/src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Constructors.jl:27 [inlined]
      From worker 3:	    [6] #algebraic_set#5797
      From worker 3:	      @ ~/ssd-data/ssd-runner-21/_work/Oscar.jl/Oscar.jl/src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Constructors.jl:50 [inlined]
      From worker 3:	    [7] algebraic_set
      From worker 3:	      @ ~/ssd-data/ssd-runner-21/_work/Oscar.jl/Oscar.jl/src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Constructors.jl:46 [inlined]
      From worker 3:	    [8] #variety#5800
      From worker 3:	      @ ~/ssd-data/ssd-runner-21/_work/Oscar.jl/Oscar.jl/src/AlgebraicGeometry/Schemes/ProjectiveVariety/Objects/Constructors.jl:52 [inlined]
      From worker 3:	    [9] surface(n::String)
      From worker 3:	      @ Oscar ~/ssd-data/ssd-runner-21/_work/Oscar.jl/Oscar.jl/src/AlgebraicGeometry/Surfaces/SurfacesP4.jl:56
      From worker 3:	   [10] veronese()
      From worker 3:	      @ Oscar ~/ssd-data/ssd-runner-21/_work/Oscar.jl/Oscar.jl/src/AlgebraicGeometry/Surfaces/SurfacesP4.jl:583
      From worker 3:	   [11] macro expansion
      From worker 3:	      @ ~/ssd-data/ssd-runner-21/_work/Oscar.jl/Oscar.jl/test/Modules/UngradedModules.jl:1335 [inlined]
      From worker 3:	   [12] macro expansion
      From worker 3:	      @ ~/ssd-data/ssd-runner-21/_work/_tool/julia/1.11.6/x64/share/julia/stdlib/v1.11/Test/src/Test.jl:1709 [inlined]
      From worker 3:	   [13] top-level scope
      From worker 3:	      @ ~/ssd-data/ssd-runner-21/_work/Oscar.jl/Oscar.jl/test/Modules/UngradedModules.jl:1334
```

Reply:
Mathematically, these objects are just the same. So I would not want to manually curate which attributes to set and which ones not.

Reply:
But there could be things in the attribute storage that are not of a mathematical nature: think e.g. of using the attribute storage to cache the `hash` of such an object. The hash value of the two objects is different (in general), so just copying the cached value may produce wrong results when later checking for equality. (I know that this is not the case right now, but we wouldn't notice the issue when someone adds an optimization like that.)

Still, maybe we need to add some way to AA to copy attributes from one object to another without relying on internals. opinions @fingolfin @thofma ?

Reply:
O.k. I suppose it is not worth the maintainance burden either way. 
Let's just forget the attributes.

Reply:
There is something I still don't understand...

> Mathematically, these objects are just the same. So I would not want to manually curate which attributes to set and which ones not.

If those objects are the same, then why are those attributes not already set for the input objects? Why does "converting" a `ProjectiveAlgebraicSet` `X` into a `ProjectiveVariety` `Z` lead to `X` suddenly getting new attributes set?

Reply:
Because the user asserts that `X` is geometrically integral, which is a property that is very expensive to check.

Reply:
OK, but then that's a special thing in this case: you assume the input is valid and therefore you deduce and store properties of the input. That's certainly something you can do, but not the norm, which is why there is a clash with the model.

In GAP we model this kind of thing with a set of helper functions: `UseFactorRelation`, `UseIsomorphismRelation`, `UseSubsetRelation`. These functions take two objects (say, two groups) as input and transfer information about them in one direction according the specified relation: "H is a isomorphic to (a subobject / a factor / all of) G". The idea then is to call this in certain places when transfer of information is useful. There is a also scheme how to conveniently specify new such "implications". And of course it is not perfect, e.g. sometimes going the other way can be useful (e.g. "if G has an infinite factor or subgroup it must be infinite itself") but in practice this is not as common (at least in the context of GAP and groups, I am not making claims about other areas of math)

Sebastian Gutsche (with Mohamed Barakat and Sebastian Posur) implemented a system which is more dynamically: once it is known that two objects are isomorphic, this is tracked, and any information learned about one is then transferred to the other. However, while this may be a win in some scenarios, it definitely isn't always, as just the effort of transferring data around all the time (which most of the time isn't needed) takes up a lot of resources.